synapse/lib/trigger.py

```diff
@@ -1,69 +1,262 @@
+                self.core.getStormQuery(rule.storm)
+            except Exception as e:
+                logger.warning('Invalid rule %r found in storage: %r.  Removing.', iden, e)
+                to_delete.append(iden)
+        for iden in to_delete:
+            self.delete(iden)
 
-        for prop, match in must[1].items():
+        self.enabled = True
 
-            valu = perm[1].get(prop)
-            if valu is None:
-                return False
+        # Re-evaluate all the events that occurred before we were enabled
+        for node, cond, info in self._deferred_events:
+            await self.fire(node, cond, info=info)
 
-            if not self._trig_match.match(valu, match):
-                return False
+        self._deferred_events.clear()
 
-        return True
+    def _load_all(self, slab):
+        db = slab.initdb(self.TRIGGERS_DB_NAME)
+        for iden, val in self.core.slab.scanByRange(b'', db=db):
+            try:
+                ruledict = s_msgpack.un(val)
+                ver = ruledict.pop('ver')
+                cond = ruledict.pop('cond')
+                user = ruledict.pop('user')
+                query = ruledict.pop('storm')
+                self._load_rule(iden, ver, cond, user, query, info=ruledict)
+            except Exception as e:
+                logger.warning('Invalid rule %r found in storage: %r', iden, e)
+                continue
 
-    def trigger(self, perm, *args, **kwargs):
-        '''
-        Fire any matching trigger functions for the given perm.
+    def _load_rule(self, iden, ver, cond, user, query, info):
+
+        rule = Triggers.Rule(ver, cond, user, query, **info)
+
+        ''' Make sure the query parses '''
+        if self.enabled:
+            self.core.getStormQuery(rule.storm)
+
+        self._rules[iden] = rule
+        if rule.prop is not None:
+            self._rule_by_prop[rule.prop].append(rule)
+        elif rule.form is not None:
+            self._rule_by_form[rule.form].append(rule)
+        else:
+            assert rule.tag
+            self._rule_by_tag[rule.tag].append(rule)
+        return rule
+
+    def list(self):
+        return [(iden, dataclasses.asdict(rule)) for iden, rule in self._rules.items()]
+
+    def mod(self, iden, query):
+        self._rules[iden].storm = query
+
+    @contextlib.contextmanager
+    def _recursion_check(self):
+        depth = RecursionDepth.get()
+        if depth > self.RECURSION_LIMIT:
+            raise s_exc.RecursionLimitHit(mesg='Hit trigger limit')
+        token = RecursionDepth.set(depth + 1)
+
+        try:
+            yield
 
-        Args:
-            perm ((str,dict)):  The perm tufo to trigger
-            *args (list):       args list to use calling the trigger function
-            **kwargs (dict):    kwargs dict to use calling the trigger function
+        finally:
+            RecursionDepth.reset(token)
 
-        Returns:
-            (None)
+    async def fire(self, node, cond, *, info):
```

Comment:
if we're not a `EventBus` or a `Base` object, do we want to be using the function `fire()` ?  Having seen it used in other code I was expecting the Trigger object to be using the observable pattern.

Reply:
ok. After trying several alternatives, it is now `run`